types/types.go

```diff
@@ -148,86 +148,66 @@ func (m *memMarker) Count(states ...AlertState) int {
 }
 
 // SetSilenced implements Marker.
-func (m *memMarker) SetSilenced(alert model.Fingerprint, version int, ids ...string) {
+func (m *memMarker) SetSilenced(alert model.Fingerprint, version int, activeIDs []string, pendingIDs []string) {
 	m.mtx.Lock()
+	defer m.mtx.Unlock()
 
 	s, found := m.m[alert]
 	if !found {
 		s = &AlertStatus{}
 		m.m[alert] = s
 	}
+	s.SilencedBy = activeIDs
+	s.pendingSilences = pendingIDs
 	s.silencesVersion = version
 
 	// If there are any silence or alert IDs associated with the
 	// fingerprint, it is suppressed. Otherwise, set it to
-	// AlertStateUnprocessed.
-	if len(ids) == 0 && len(s.InhibitedBy) == 0 {
-		m.mtx.Unlock()
-		m.SetActive(alert)
+	// AlertStateActive.
+	if len(activeIDs) == 0 && len(s.InhibitedBy) == 0 {
+		s.State = AlertStateActive
 		return
 	}
 
 	s.State = AlertStateSuppressed
-	s.SilencedBy = ids
-
-	m.mtx.Unlock()
 }
 
 // SetInhibited implements Marker.
 func (m *memMarker) SetInhibited(alert model.Fingerprint, ids ...string) {
 	m.mtx.Lock()
+	defer m.mtx.Unlock()
 
 	s, found := m.m[alert]
 	if !found {
 		s = &AlertStatus{}
 		m.m[alert] = s
 	}
+	s.InhibitedBy = ids
 
 	// If there are any silence or alert IDs associated with the
 	// fingerprint, it is suppressed. Otherwise, set it to
-	// AlertStateUnprocessed.
+	// AlertStateActive.
 	if len(ids) == 0 && len(s.SilencedBy) == 0 {
-		m.mtx.Unlock()
-		m.SetActive(alert)
+		s.State = AlertStateActive
 		return
 	}
 
 	s.State = AlertStateSuppressed
-	s.InhibitedBy = ids
-
-	m.mtx.Unlock()
-}
-
-// SetActive implements Marker.
-func (m *memMarker) SetActive(alert model.Fingerprint) {
-	m.mtx.Lock()
-	defer m.mtx.Unlock()
-
-	s, found := m.m[alert]
-	if !found {
-		s = &AlertStatus{}
-		m.m[alert] = s
-	}
-
-	s.State = AlertStateActive
-	s.SilencedBy = []string{}
-	s.InhibitedBy = []string{}
 }
 
 // Status implements Marker.
 func (m *memMarker) Status(alert model.Fingerprint) AlertStatus {
 	m.mtx.RLock()
 	defer m.mtx.RUnlock()
 
-	s, found := m.m[alert]
-	if !found {
-		s = &AlertStatus{
-			State:       AlertStateUnprocessed,
-			SilencedBy:  []string{},
-			InhibitedBy: []string{},
-		}
+	if s, found := m.m[alert]; found {
+		return *s
+	}
+	return AlertStatus{
+		State:       AlertStateUnprocessed,
+		SilencedBy:  []string{},
+		InhibitedBy: []string{},
```

Comment:
Yeah, generally, it's good Go design to allow `nil` instead of an empty slice. That's why I didn't set `pendingSilences`. But since `SilencedBy` and `InhibitedBy` are exported, I wasn't sure if I break any external assumptions, in particular because `AlertStatus` is marshaled as JSON, and I didn't want to take bets (or investigate in detail) if `nil` is marshaled in a different way as an empty string.

Reply:
Makes sense.